File: 02.xtrem-t.d.d/src/main/java/money_problem/domain/Portfolio.java
package money_problem.domain;

import java.util.ArrayList;
import java.util.List;

public class Portfolio {

    private final List<Money> lines = new ArrayList<>();

    private static String formatPortfolio(Currency currency, double sum) {
        return sum + " " + currency;
    }

    private Money convertToCurrency(Currency currency, Money money, Bank bank) {
        try {
            return bank.convert(money, currency);
        } catch (MissingExchangeRateException e) {
            throw new RuntimeException(e);
        }
    }

    public String evaluate(Bank bank, Currency currency) {
        double sum = lines
                .stream()
                .map(money -> convertToCurrency(currency, money, bank).amount()
                ).reduce(0.0, Double::sum);
        return formatPortfolio(currency, sum);
    }

    public void addMoney(Money money) {
        lines.add(money);
    }
}

File: 02.xtrem-t.d.d/src/main/kotlin/money_problem/core/Portfolio.kt
package money_problem.core

import arrow.core.Either
import arrow.core.left
import arrow.core.right

typealias ConversionResults = List<Either<MissingExchangeRateError, Money>>

class Portfolio private constructor(private val lines: List<Money>) {
    constructor() : this(emptyList())

    fun addMoney(money: Money): Portfolio = Portfolio(lines + money)

    private fun convertAllMoneys(bank: Bank, toCurrency: Currency): List<Either<MissingExchangeRateError, Money>> =
        lines.map { money: Money -> bank.convert(money, toCurrency) }

    fun evaluate(bank: Bank, to: Currency): Either<MissingExchangeRatesError, Money> =
        convertAllMoneys(bank, to)
            .let {
                when {
                    it.containsFailure() -> toFailure(it).left()
                    else -> sumConvertedMoney(it, to).right()
                }
            }

    private fun ConversionResults.containsFailure(): Boolean = any { result -> result.isLeft() }

    private fun toFailure(convertedMoneys: ConversionResults): MissingExchangeRatesError =
        MissingExchangeRatesError(
            convertedMoneys
                .filter { result -> result.isLeft() }
                .mapNotNull { result -> result.swap().orNull() }
        )

    private fun sumConvertedMoney(convertedMoneys: ConversionResults, toCurrency: Currency): Money =
        Money(
            convertedMoneys
                .mapNotNull { it.orNull() }
                .sumOf { it.amount }, toCurrency
        )
}